pkg/kp/dsstore/consul_store.go

```diff
@@ -0,0 +1,207 @@
+type CASError string
+
+func (e CASError) Error() string {
+	return fmt.Sprintf("Could not check-and-set key %q", string(e))
+}
+
+// Make sure functions declared in the Store interface have the proper contract
+var _ Store = &consulStore{}
+
+func NewConsul(client *api.Client, retries int, logger *logging.Logger) *consulStore {
+	return &consulStore{
+		retries:    retries,
+		applicator: labels.NewConsulApplicator(client, retries),
+		kv:         client.KV(),
+		logger:     *logger,
+	}
+}
+
+func (s *consulStore) Create(
+	manifest pods.Manifest,
+	minHealth int,
+	name fields.ClusterName,
+	nodeSelector klabels.Selector,
+	podID types.PodID,
+) (fields.DaemonSet, error) {
+	if err := checkManifestPodID(podID, manifest); err != nil {
+		return fields.DaemonSet{}, err
+	}
+
+	ds, err := s.innerCreate(manifest, minHealth, name, nodeSelector, podID)
+
+	// TODO: measure whether retries are is important in practice
+	for i := 0; i < s.retries; i++ {
+		if _, ok := err.(CASError); ok {
+			ds, err = s.innerCreate(manifest, minHealth, name, nodeSelector, podID)
+		} else {
+			break
+		}
+	}
+	if err != nil {
+		return fields.DaemonSet{}, err
+	}
+	return ds, nil
+}
+
+// these parts of Create may require a retry
+func (s *consulStore) innerCreate(
+	manifest pods.Manifest,
+	minHealth int,
+	name fields.ClusterName,
+	nodeSelector klabels.Selector,
+	podID types.PodID,
+) (fields.DaemonSet, error) {
+	id := fields.ID(uuid.New())
+	dsPath, err := s.dsPath(id)
+	if err != nil {
+		return fields.DaemonSet{}, err
+	}
+	ds := fields.DaemonSet{
+		ID:           id,
+		Disabled:     false,
+		Manifest:     manifest,
+		MinHealth:    minHealth,
+		Name:         name,
+		NodeSelector: nodeSelector,
+		PodID:        podID,
+	}
+	// Marshals ds into []bytes using overloaded MarshalJSON
+	rawDS, err := json.Marshal(ds)
+	if err != nil {
+		return fields.DaemonSet{}, util.Errorf("Could not marshal DS as json: %s", err)
+	}
+
+	success, _, err := s.kv.CAS(&api.KVPair{
+		Key:         dsPath,
+		Value:       rawDS,
+		ModifyIndex: 0,
+	}, nil)
+
+	if err != nil {
+		return fields.DaemonSet{}, consulutil.NewKVError("cas", dsPath, err)
+	}
+	if !success {
+		return fields.DaemonSet{}, CASError(dsPath)
+	}
+	return ds, nil
+}
+
+func (s *consulStore) Get(id fields.ID) (fields.DaemonSet, *api.QueryMeta, error) {
+	dsPath, err := s.dsPath(id)
+	var metadata *api.QueryMeta
+
+	if err != nil {
+		return fields.DaemonSet{}, metadata, err
+	}
+
+	kvp, metadata, err := s.kv.Get(dsPath, nil)
+	if err != nil {
+		return fields.DaemonSet{}, metadata, consulutil.NewKVError("get", dsPath, err)
+	}
+	if metadata == nil {
+		// no metadata returned
+		return fields.DaemonSet{}, metadata, errors.New("No metadata found")
+	}
+	if kvp == nil {
+		// ID didn't exist
+		return fields.DaemonSet{}, metadata, NoDaemonSet
+	}
+
+	ds, err := kvpToDS(kvp)
+	if err != nil {
+		return fields.DaemonSet{}, metadata, err
+	}
+	return ds, metadata, err
```

Comment:
you don't need the `err` check above the way this is written, or you can return `nil` for the error here